### PR TITLE
Implement Validation for Filtering

### LINKED ISSUE #72: Implement validation for filtering
Prereq: https://github.com/sandboxnu/cooper/issues/71

Now that we can accept an object that will filter our results, we want to validate this object to make sure we're filtering by the right things. 

For now, we only want to filter by work model and location. 

Acceptance criteria:
- Any object that tries to filter by something that is not work model and location is rejected with the appropriate status code  when endpoint is queried
- Any object that tries to filter by only work model and/or location returns correct data when endpoint is quiered

### PR DESCRIPTION
**Pull request by gpalmer27**
## Description

Makes sure that the workTerm and workEnvironment that is inputted is valid and gives an error otherwise.

## Motivation and Context

Closes #72

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have moved the ticket to "In Review"
- [x] I have added reviewers to this PR
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: apps/web/src/app/(pages)/(dashboard)/roles/page.tsx [hunk 1]
```diff
@@ -22,6 +24,24 @@ export default function Roles({
     workEnvironment?: WorkEnvironmentType;
   };
 }) {
+  const [error, setError] = useState<string | undefined>(undefined);
+
+  useEffect(() => {
+    const isValidTerm = searchParams?.workTerm
```
Comment (joshw2048) [inline apps/web/src/app/(pages)/(dashboard)/roles/page.tsx:30]: i'm fairly certain we don't need to check for random invalid search params, like if someone puts josh=best-pl-ever, we can safely ignore

File: apps/web/src/app/(pages)/(dashboard)/roles/page.tsx [hunk 2]
```diff
@@ -33,10 +53,57 @@ export default function Roles({
...
 
+  const [showAlert, setShowAlert] = useState(true);
+
   return (
     <>
       <SearchFilter />
       {/* TODO: Loading animations */}
+      {error && showAlert && (
+        <div
+          id="bad-input"
```
Comment (joshw2048) [inline apps/web/src/app/(pages)/(dashboard)/roles/page.tsx:64]: i think we should make this a component so we can reuse this toast pop up in other places, @RishikeshNK thoughts?

**Review by joshw2048 — DISMISSED** [review]

File: apps/web/src/app/(pages)/(dashboard)/roles/page.tsx (hunk 1 above)
Comment (joshw2048) [reply apps/web/src/app/(pages)/(dashboard)/roles/page.tsx:30]: just a comment, not a change

File: apps/web/src/app/(pages)/(dashboard)/roles/page.tsx [hunk 3]
```diff
@@ -1,10 +1,12 @@
 "use client";
 
-import { useState } from "react";
+import { useEffect, useState } from "react";
 
-import type {
+import {
```
Comment (RishikeshNK) [inline apps/web/src/app/(pages)/(dashboard)/roles/page.tsx:5]: You can fix this warning by doing something like this:

```typescript
import type {
  ReviewType,
  WorkEnvironmentType,
  WorkTermType,
} from "@cooper/db/schema";
import { WorkEnvironment, WorkTerm } from "@cooper/db/schema";
```

File: apps/web/src/app/(pages)/(dashboard)/roles/page.tsx [hunk 4]
```diff
@@ -22,6 +24,24 @@ export default function Roles({
     workEnvironment?: WorkEnvironmentType;
   };
 }) {
+  const [error, setError] = useState<string | undefined>(undefined);
+
+  useEffect(() => {
```
Comment (RishikeshNK) [inline apps/web/src/app/(pages)/(dashboard)/roles/page.tsx:29]: We _could_ get away without using a `useEffect` or possibly a `useState` for tracking the errors!

We can define a `RolesSearchParam` schema in the `validators` package using Zod. Could be something like the following:

```typescript
const SearchParamsSchema = z.object({
  cycle: z.nativeEnum(WorkTerm).optional(),
  term: z.nativeEnum(WorkEnvironment).optional(),
});
```

We can share this object between the web app and the `list` endpoint for reviews. For validation, we can do something like the following:

```typescript
const validationResult = SearchParamsSchema.safeParse(searchParams);

const reviews = api.review.list.useQuery({
  options: validationResult.success ? validationResult.data : {},
});
```

At the same time, we can also use the same `success` property to display the error modal / toast:

```typescript
{!validationResult.success && (
        <ErrorAlert message="Invalid search parameters." />
      )}
```

If you want more specific error messages, you can modify the Zod schema like this:

```typescript
const SearchParamsSchema = z.object({
  cycle: z
    .nativeEnum(WorkTerm, {
      message: "Invalid cycle type",
    })
    .optional(),
  term: z
    .nativeEnum(WorkEnvironment, {
      message: "Invalid term type",
    })
    .optional(),
});
```

and get the errors using `validationResult.error`.

File: apps/web/src/app/(pages)/(dashboard)/roles/page.tsx (hunk 2 above)
Comment (RishikeshNK) [reply apps/web/src/app/(pages)/(dashboard)/roles/page.tsx:64]: Yup! I agree. We can move this to a separate component that takes in the `message` as a prop.

File: apps/web/src/app/(pages)/(dashboard)/roles/page.tsx [hunk 5]
```diff
@@ -33,10 +53,57 @@ export default function Roles({
...
   );
 
+  const [showAlert, setShowAlert] = useState(true);
+
   return (
     <>
       <SearchFilter />
       {/* TODO: Loading animations */}
+      {error && showAlert && (
+        <div
```
Comment (RishikeshNK) [inline apps/web/src/app/(pages)/(dashboard)/roles/page.tsx:63]: Totally optional but I'd recommend taking a look at [shadcn/ui](https://ui.shadcn.com/). Almost all of our reusable components are built on top of shadcn. They do have a [toast](https://ui.shadcn.com/docs/components/toast) that you can use! 

If you're interested, check out [this](https://cooper-docs.vercel.app/docs/web/ui) docs page. You can add the component from the root folder using `pnpm ui-add toast` and use that component here.

**Review by RishikeshNK — CHANGES_REQUESTED** [review]
Looks good! Added a few suggestions to make use of our existing validation library (Zod) which should eliminate the need for the `useEffect` and _maybe_ using a shadcn component for the toast / modal.

**Review by RishikeshNK — APPROVED** [review]
FIRE 🪨

**Pull request merged**